### DISCUSSION TITLE
Parsing conversion build error after migrating to Swift 5.8

### DISCUSSION THREAD

**Discussion by chwo** [pointfreeco/swift-url-routing#79] [Q&A] [answered]
Hello!

I am having trouble migrating the following code to the new result builders syntax in Swift 5.8. The code uses a conversion in the parsing block.

Minimal code sample:

```swift
enum AppRoute: Equatable {
    case route(
        value: String? = nil,
        subRoute: SubRoute? = nil
    )
}
enum SubRoute: String {
    case sub1
    case sub2
}

// let router: some Router<AppRoute> = OneOf {
let router = OneOf {
    Route(.case(AppRoute.route)) {
        Parse(routeConversion) {
            OneOf {
                Path { "path1" }.map { SubRoute.sub1 }
                Path { "path2" }.map { SubRoute.sub2 }
            }
            Always<URLRequestData, String?>(nil) // simplified for demo purpose
        }
    }
    // .. other route cases  
}

let routeConversion: AnyConversion<(SubRoute?, String?), (String?, SubRoute?)> = .convert(
    apply: { subRoute, value in
        (value, subRoute)
    },
    unapply: { value, subRoute in
        (subRoute, value)
    }
)
```

The code did work fine in Xcode 14.2 with `swift-url-routing` version `0.4.0`.

But it doesn't work in Xcode 14.3 with `swift-url-routing` version `0.5.0`. After specifying the opaque result type with `some Router<AppRoute>` (based on https://github.com/pointfreeco/swift-parsing/discussions/290) it results in the following build error:

`Initializer 'init(_:with:)' requires the types '(String?, SubRoute?)' and '(String?, SubRoute)' be equivalent`

`Where 'C.Input' = '(String?, SubRoute?)', 'P.Output' = '(String?, SubRoute)' (DerivedData/../SourcePackages/checkouts/swift-url-routing/Sources/URLRouting/Route.swift)`

However changing the `SubRoute` in the conversion to optional doesn't seem right because this would require to adjust the parsing logic as well which should remain the same (so that this parsing block only matches the route, if one of the subpaths are matches).

What is the right way to fix this error? Is there another way to explicitly specify the types in the parsing block when using a conversion?

**Comment by chwo** [comment]
@stephencelis You mentioned this in the [swift-parsing release announcement thread]( https://github.com/pointfreeco/swift-parsing/discussions/290#discussioncomment-5513865):

> Swift result builders in 5.8 no longer support bidirectional type inference, so you need to be more explicit with types sometimes.

Could this also be the case here? Do you have a tip how to specify the type more explicitly or do you have another idea how to fix this?

**Comment by kamcma** [comment]
I wonder if your conversion trouble isn't a downstream consequence of incorrectly modeled routes. You've modeled `AppRoute`'s sole case to have an optional associated value of `SubRoute`. But look through [the modeled routes in the pointfreeco repository](https://github.com/pointfreeco/pointfreeco/blob/06152b56454e08f4acdc206054e03d0389a883ed/Sources/PointFreeRouter/Routes.swift). Sub routes are not modeled as optional associated values, they are modeled as _cases of_ parent routes. There are some optional values, but for things like optional query parameters, body data, etc, not any sub routes.

I tried to have a go at solving your compiler error, but because of the oddly modeled routes, I can't even quite make out what routes you intend to model. I feel like if you clarified your route modeling, you wouldn't need to do that ad-hoc conversion between `SubRoute` and `String?` at all.

**Reply by chwo** [reply, marked as answer]
@kamcma Thank you for looking into this. Regarding the sub route behavior, I just named it `SubRoute` for demo purposes, it is actually more a configuration value derived from the path or query parameters.

---

Fortunately, I figured out the solution myself by having a closer look at the error message again:

`Initializer 'init(_:with:)' requires the types '(String?, SubRoute?)' and '(String?, SubRoute)' be equivalent`

- The first sub route (`SubRoute?`) is optional because it is modelled as optional in the enum
- The compiler complains about the second sub route (`SubRoute`) being non optional because this is inferred from the parsing block:

```swift
OneOf {
    Path { "path1" }.map { SubRoute.sub1 }
    Path { "path2" }.map { SubRoute.sub2 }
}
```

So to fix this, the solution is to explicitly write the parsing block as optional sub route:

```swift
OneOf {
    Path { "path1" }.map { Optional(SubRoute.sub1) }
    Path { "path2" }.map { Optional(SubRoute.sub2) }
}
```

**Reply by kamcma** [reply]
Again, hard for me to say because it's a little hard to infer what actual (ie, string) routes are intended to be modeled from the route types shown (eg, I can't tell how `value` is intended to be used), but I think the library happy path has you model your routes/router a little more like this:
```swift
enum AppRoute {
  case index
  case subRoute(SubRoute)
}

enum SubRoute {
  case sub1
  case sub2
}

let router = OneOf {
  Route(.case(AppRoute.index))

  Route(.case(AppRoute.subRoute)) {
    OneOf {
      Route(.case(SubRoute.sub1)) {
        Path { "path1" }
      }

      Route(.case(SubRoute.sub2)) {
        Path { "path2" }
      }
    }
  }
}
```
That requires no custom conversions, and models: `/`, `/path1`, and `/path2`. Note that the root app route is modeled as separate case of the `AppRoute` enum, not as the nil case of an optional associated value in the `subRoute` case.
